src/CoreWf/AsyncTaskCodeActivity.cs

```diff
@@ -0,0 +1,154 @@
+﻿using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Activities
+{
+    public abstract class AsyncTaskCodeActivity : AsyncCodeActivity, IAsyncTaskCodeActivity
+    {
+        protected sealed override IAsyncResult BeginExecute(AsyncCodeActivityContext context, AsyncCallback callback, object state)
+        {
+            var cts = new CancellationTokenSource();
+            context.UserState = cts;
+
+            var task = ExecuteAsync(context, cts.Token);
+
+            // VoidResult allows us to simulate a void return type.
+            var tcs = new TaskCompletionSource<VoidResult>(state);
```

Comment:
.NET 5 adds a new non-generic `TaskCompletionSource` to handle exactly this situation. However, the type is not backwards-compatible. I originally had an ifdef to switch between the generic and non-generic version based on the targeted .NET version, however I felt this added unnecessary complexity and opted to just use the generic version.